### PR TITLE
Improve the efficiency of the S3 storage provider

### PR DESCRIPTION
**Pull request by richvdh**
there are a few separate things here, which I hope will mean that (a) we use less CPU, and (b) what CPU we do use gets traced to the requests that cause it rather than getting lost down the sofa.

### CODE REVIEW

File: s3_storage_provider.py [hunk 1]
```diff
@@ -76,41 +77,47 @@ def __init__(self, hs, config):
...
+            self._s3_client = s3 = b3_session.client("s3", **self.api_kwargs)
+        return s3
+
     def store_file(self, path, file_info):
         """See StorageProvider.store_file"""
 
         def _store_file():
-            session = boto3.session.Session()
-            session.resource("s3", **self.api_kwargs).Bucket(self.bucket).upload_file(
+            self._get_s3_client().upload_file(
```
Comment (richvdh) [inline s3_storage_provider.py]: I'm struggling to find good docs on the API for boto, but according to https://boto3.amazonaws.com/v1/documentation/api/latest/guide/s3-uploading-files.html I think this should work?

**Comment by erikjohnston** [comment]
I'm failing to see anywhere in the link in your commit that boto clients are thread safe? Also, was that a change that you think will speed things up at all? I'd have thought using thread local storage meant we'd rarely have to create a new session.

**Comment by richvdh** [comment]
> I'm failing to see anywhere in the link in your commit that boto clients are thread safe? 

At the bottom of https://boto3.amazonaws.com/v1/documentation/api/latest/guide/resources.html:

image

>  Also, was that a change that you think will speed things up at all?   Also, was that a change that you think will speed things up at all? 

I was less worried about instantiating *clients* for each thread (which as you say, were being stored in threadlocals so relatively low-overhead); using a single client was really just a simplification.

The problem was that the *upload* path was creating a new *resource* for each upload, and resources are expensive to create.

**Comment by erikjohnston** [comment]
Aha, thanks.

**Review by erikjohnston — APPROVED** [review]

**Review by erikjohnston — APPROVED** [review]

**Pull request merged**